### DISCUSSION TITLE
Correct syntax for driver.executeScript's 'windows: click'?

### DISCUSSION THREAD

**Discussion by MathieuQA** [webdriverio/webdriverio#12918] [Q&A]
Hey all! I've been trying to do some desktop automation with the following: Node.js, WDIO, Appium, WinAppDriver. Full devDependencies at the end.

In the hopes of achieving double clicks and drag and drop, I've been using lines like driver.executeScript("windows: click", elementId: "e.id", times: 2). Unfortunately, I'm getting this error: 
InvalidArgumentError: Either element identifier or absolute coordinates must be provided

Looking at my appium server logs, I find this line: 
[WindowsDriver@42c3 (2db8c7c9)] Calling AppiumDriver.execute() with args: ["windows: click",[[{"x":62,"y":292,"times":2}]],"2db8c7c9-c45e-47e6-b9bc-19fa7d32cacd"]

Specifically, it seems that the line the appium server receives has a weird amount of brackets? I posted on the appium forums and they seemed to find that the most likely source of the issue and recommended I ask around here.

What's the proper syntax to write this method? With that error, how could I go ahead and try solving it?

Thank you in advance :)

“devDependencies”: {
“@babel/core”: “^7.24.1”,
“@babel/preset-env”: “^7.24.1”,
“@babel/preset-typescript”: “^7.24.1”,
“@types/jest”: “^29.5.12”,
“@types/selenium-webdriver”: “^4.1.22”,
“babel-jest”: “^29.7.0”,
“jest”: “^29.7.0”,
“jest-junit-reporter”: “1.1.0”,
“selenium-appium”: “^1.0.2”,
“selenium-webdriver”: “^4.18.1”,
“typescript”: “^5.4.2”,
“wdio-ywinappdriver-service”: “^0.2.54”,
“winappdriver”: “^0.0.7”
}

**Comment by christian-bromann** [comment]
@MathieuQA thanks for raising this discussion. I recommend to reach out to the Appium team. WebdriverIO only sends the command as mentioned above: `driver.executeScript("windows: click", elementId: "e.id", times: 2)` and has no influence how Appium calculates the coordinates.

Let me know if you have further questions.

**Reply by MathieuQA** [reply]
Hi! Thank you for your response Christian. I did open this with Appium, on their forums here: https://discuss.appium.io/t/best-ways-to-double-click-and-drag-and-drop-on-desktop/42536/12

Mykola, which to my understanding is an appium dev, asked me to contact WDIO. So now I'm a bit unsure what to do now. 

Could you confirm if the syntax I use is correct? Is the commantd I'm showing supposed to be working? I'm sorry to bother if this isn't in your court.

**Reply by christian-bromann** [reply]
Can you share a piece of code that is failing, including some logs?

**Reply by MathieuQA** [reply]
Sure! Here's a short test script I used to test that:

```
await By2.nativeXpath('//Pane/Group/Group/Group/Group/Group/Button[1]').click();
await By2.nativeName('Projects').click();
const fileId = getElementId('Contrat - 1 Page - FR.pdf')
await driver.executeScript('windows: click', {"elementId": fileId, "times": 2});

```
getElementId is the following method that converts a nativeName into an ID:
```
async function getElementId(elementName: string): Promise<string> {
  const element: WebElement = await driver.findElement(By2.nativeName(elementName));
  const elementId: string = await element.getId();
  return elementId;
}

```

I've tried with a few ways, either XY coordinates or element ID. The following is some appium server logs that appeared back when I used XY coordinates, but it's pretty much the exact same with element ID.

```
[WindowsDriver@42c3 (2db8c7c9)] Calling AppiumDriver.execute() with args: ["windows: click",[[{"x":62,"y":292,"times":2}]],"2db8c7c9-c45e-47e6-b9bc-19fa7d32cacd"]
[WindowsDriver@42c3 (2db8c7c9)] Executing extension command 'windows: click'
[WindowsDriver@42c3 (2db8c7c9)] Encountered internal error running command: InvalidArgumentError: Either element identifier or absolute coordinates must be provided
[WindowsDriver@42c3 (2db8c7c9)]     at WindowsDriver.toAbsoluteCoordinates (C:\Users\mathieu.chouinardlav\.appium\node_modules\appium-windows-driver\lib\commands\gestures.js:64:11)
[WindowsDriver@42c3 (2db8c7c9)]     at WindowsDriver.windowsClick (C:\Users\mathieu.chouinardlav\.appium\node_modules\appium-windows-driver\lib\commands\gestures.js:300:72)
[WindowsDriver@42c3 (2db8c7c9)]     at WindowsDriver.executeWindowsCommand (C:\Users\mathieu.chouinardlav\.appium\node_modules\appium-windows-driver\lib\commands\execute.js:46:57)
[WindowsDriver@42c3 (2db8c7c9)]     at WindowsDriver.execute (C:\Users\mathieu.chouinardlav\.appium\node_modules\appium-windows-driver\lib\commands\execute.js:33:23)
[WindowsDriver@42c3 (2db8c7c9)]     at commandExecutor (C:\Users\mathieu.chouinardlav\.appium\node_modules\@appium\base-driver\lib\basedriver\driver.ts:106:18)
[WindowsDriver@42c3 (2db8c7c9)]     at C:\Users\mathieu.chouinardlav\.appium\node_modules\async-lock\lib\index.js:171:12
[WindowsDriver@42c3 (2db8c7c9)]     at AsyncLock._promiseTry (C:\Users\mathieu.chouinardlav\.appium\node_modules\async-lock\lib\index.js:306:31)
[WindowsDriver@42c3 (2db8c7c9)]     at exec (C:\Users\mathieu.chouinardlav\.appium\node_modules\async-lock\lib\index.js:170:9)
[WindowsDriver@42c3 (2db8c7c9)]     at AsyncLock.acquire (C:\Users\mathieu.chouinardlav\.appium\node_modules\async-lock\lib\index.js:189:3)
[WindowsDriver@42c3 (2db8c7c9)]     at WindowsDriver.executeCommand (C:\Users\mathieu.chouinardlav\.appium\node_modules\@appium\base-driver\lib\basedriver\driver.ts:122:39)
[WindowsDriver@42c3 (2db8c7c9)]     at processTicksAndRejections (node:internal/process/task_queues:95:5)
[WindowsDriver@42c3 (2db8c7c9)]     at defaultBehavior (C:\Users\mathieu.chouinardlav\AppData\Roaming\npm\node_modules\appium\lib\appium.js:1109:14)
[WindowsDriver@42c3 (2db8c7c9)]     at AppiumDriver.executeWrappedCommand (C:\Users\mathieu.chouinardlav\AppData\Roaming\npm\node_modules\appium\lib\appium.js:1215:16)
[WindowsDriver@42c3 (2db8c7c9)]     at AppiumDriver.executeCommand (C:\Users\mathieu.chouinardlav\AppData\Roaming\npm\node_modules\appium\lib\appium.js:1121:17)
[WindowsDriver@42c3 (2db8c7c9)]     at asyncHandler (C:\Users\mathieu.chouinardlav\AppData\Roaming\npm\node_modules\appium\node_modules\@appium\base-driver\lib\protocol\protocol.js:393:19)
```

Mykola mentioned the following part as weird: 
[WindowsDriver@42c3 (2db8c7c9)] Calling AppiumDriver.execute() with args: ["windows: click",**[[{"x":62,"y":292,"times":2}]]**,"2db8c7c9-c45e-47e6-b9bc-19fa7d32cacd"]
Specifically, the amount of brackets, like it's multiple arrays or something? It does seem abnormal, because I don't include those brackets at all. [[{"elementId":"814221c6-4b46-41bc-9312-82ccf6f494c0","times":2}]] is how it looks like when I use elementId, despite the actual command looking like {"elementId": fileId, "times": 2});

**Reply by christian-bromann** [reply]
Can you this:

```ts
await driver.execute('windows: click', [{"elementId": fileId, "times": 2}]);
```

**Reply by MathieuQA** [reply]
I'm getting "Argument of type 'string' is not assignable to parameter of type 'Command'.ts(2345)" for 'windows:click'. I tried changing execute to executeScript, but same issue as before despite the changed syntax on the arguments.

**Reply by christian-bromann** [reply]
So `await driver.executeScript('windows: click', [{"elementId": fileId, "times": 2}]);` fails as well?

**Reply by christian-bromann** [reply]
Can you provide the WDIO logs with request information?

**Reply by MathieuQA** [reply]
Yes, I get the following:
Appium server logs:
```
[WindowsDriver@4794 (b929bacf)] Calling AppiumDriver.execute() with args: ["windows: click",[[[{"elementId":"b8654255-1c5a-4cc1-bdf7-8e1f1150a9c2","times":2}]]],"b929bacf-df3f-4946-9849-dcb10ab40cd2"]      
[WindowsDriver@4794 (b929bacf)] Executing extension command 'windows: click'
[WindowsDriver@4794 (b929bacf)] Encountered internal error running command: InvalidArgumentError: Either element identifier or absolute coordinates must be provided
```

And the logs from the command prompt I launch it from:
InvalidArgumentError: Either element identifier or absolute coordinates must be provided

      64 |         const fileId = getElementId('Contrat - 1 Page - FR.pdf')
      65 |         await console.log(fileId);
    > 66 |         await driver.executeScript('windows: click', [{"elementId": fileId, "times": 2}]);
         |         ^

**Reply by MathieuQA** [reply]
I'm not entirely sure where I can find the WDIO logs, how can I get that?

**Reply by christian-bromann** [reply]
If you have `outputDir` set you find the logs in that directory, if not, in your terminal output.

**Reply by MathieuQA** [reply]
I don't have outputDir set, so here's the full output in my terminal

PS C:\Users\mathieu.chouinardlav\git\cod4-test-automation> npm run jest Demonstration.test.ts

> cod4-test-automation@1.0.0 jest
> jest --maxWorkers=1 Demonstration.test.ts

  console.log
    Promise { <pending> }

      at Object.log (__tests__/Demonstration.test.ts:65:23)

 FAIL  __tests__/Demonstration.test.ts (22.143 s)
  Tests
    × Demo (18866 ms)

  ● Tests › Demo

    InvalidArgumentError: Either element identifier or absolute coordinates must be provided

      64 |         const fileId = getElementId('Contrat - 1 Page - FR.pdf')
      65 |         await console.log(fileId);
    > 66 |         await driver.executeScript('windows: click', [{"elementId": fileId, "times": 2}]);
         |         ^
      67 |         //await driver.executeScript("windows: click", {"elementId": fileId,"times":2})
      68 |         await driver.sleep(10000);
      69 |         //"//Pane/Group/Group/Group/Group/Group/Group[3]/Group/Text[4]" "elementId": "Contrat - 1 Page - FR.pdf" elementLocated["62"], "y": elementLocated["292"]

      at Object.throwDecodedError (node_modules/selenium-webdriver/lib/error.js:521:15)
      at parseHttpResponse (node_modules/selenium-webdriver/lib/http.js:510:13)
      at Executor.execute (node_modules/selenium-webdriver/lib/http.js:443:28)
      at WebDriver.execute (node_modules/selenium-webdriver/lib/webdriver.js:740:17)
      at Object.<anonymous> (__tests__/Demonstration.test.ts:66:9)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        22.254 s, estimated 55 s
Ran all test suites matching /Demonstration.test.ts/i.

**Reply by BorisOsipov** [reply]
It seems you have to update driver https://github.com/appium/appium-windows-driver/issues/183

**Reply by MathieuQA** [reply]
That issue was recommended to me by Mykola when I contacted Appium; my driver is up to date currently. I did find it weird that the same error was resolved by Mykola with an update. I'm not sure what exact version that person was running. I uninstalled and reinstalled my driver just in case, but encountered the same issue.

**Reply by MathieuQA** [reply]
I see the issue has been closed as resolved originally, before any of our discussions. Is it safe to assume this is still ongoing? Should this be re-opened? Apologies, I'm not very familiar with github issues.  I don't think I've any ways to re-open this since it's currently ongoing.

**Reply by christian-bromann** [reply]
@MathieuQA your error stack `node_modules/selenium-webdriver/lib/error.js` suggests that you are using Selenium and __not__ WebdriverIO. Please report this issue in the Selenium project.

**Reply by MathieuQA** [reply]
Oh. The Appium client list only lists Nightwatch and WebDriverIO, so I've been told to come here. Should I be making an issue on the main selenium project, or on selenium-appium? As you can tell, I'm no experienced dev, feeling a bit lost through all this. Sorry to waste your time.

**Reply by christian-bromann** [reply]
You seem to be using Nightwatch because it uses Selenium under the hood. So the Nightwatch project would be the best place to report the issue.

> As you can tell, I'm no experienced dev, feeling a bit lost through all this.

No worries, we are here to help navigate.

**Reply by MathieuQA** [reply]
I've never heard of Nightwatch, but it's not entirely impossible I've installed it through a guide and not looked back. However, Nightwatch doesn't seem to permit any form of desktop automation and seemingly has a focus on browser automation; I'm automating a flutter based desktop application. Furthermore, searching the term "nightwatch" in my project doesn't seem to return anything. To me, it seems like I'm using Selenium itself, but it's not even listed as a client on the appium website, so something's weird. Am I potentially just missing a client? That makes no sense, right? I've been able to do things like clicking around and opening the app, so I have a full setup... right?

One thing I'd like to see is if I can just replace whatever it is I currently have with WDIO. I've no attachment to my current setup, I really just want to find something that functions for all my needs; flutter based desktop application (windows + mac) automation that can click, double click, drag and drop, and interact with other windows. I only seem to be unable to do the last three currently (but I'm pretty sure the last one is possible..). If switching to WDIO could allow me to do those things, I would love to.
